File: card/src/clients/stan.rs
use crate::clients::event_store::EventStore;
use crate::db;
use crate::workers::Workers;
use futures::{executor::block_on, StreamExt};
use ratsio::{RatsioError, StanClient, StanOptions};
use std::sync::Arc;

pub struct Stan {
    pub client: Arc<StanClient>,
}

impl Default for Stan {
    fn default() -> Self {
        Stan {
            client: block_on(init()).unwrap(),
        }
    }
}

pub async fn init() -> Result<Arc<StanClient>, RatsioError> {
    let client_id = "card".to_string();
    let stan_server_url = "localhost:4222";
    let cluster_name = "melting-pot";
    let opts = StanOptions::with_options(stan_server_url, cluster_name, &client_id[..]);
    let sc = StanClient::from_options(opts).await?;
    Ok(sc)
}

impl Stan {
    pub async fn subscribes(&self, rocket: rocket::Rocket) -> Result<(), RatsioError> {
        let event_workers = Workers {
            event_store_client: EventStore::default().client,
        };

        let user_created_subject = "user_created";
        let queue_group = None;
        let durable_name = Some("durable-card");

        let (sid, mut subscription) = self
            .client
            .subscribe(user_created_subject, queue_group, durable_name)
            .await?;

        println!("subscribed sid: {:?}", sid);

        tokio::spawn(async move {
            while let Some(message) = subscription.next().await {
                let payload = String::from_utf8_lossy(&message.payload);
                println!(
                    "Received message subject: {:?}, payload: {:?}",
                    &message.subject, payload,
                );

                let conn = db::DbConn::get_one(&rocket).unwrap();
                event_workers
                    .clone()
                    .handle_user_created(payload.to_string(), conn)
                    .await;
            }
        });
        Ok(())
    }
}

File: wallet/internal/clients/stan.go
package clients

import (
	"log"

	"melting_pot/wallet/internal/workers"

	stan "github.com/nats-io/stan.go"
)

type Stan struct {
	client stan.Conn
}

func (s *Stan) Default() {
	clusterID := "melting-pot"
	clientID := "wallet"
	sc, err := stan.Connect(clusterID, clientID)
	if err != nil {
		log.Printf("Failed to connect to nats streaming server, err: %v", err)
	}
	s.client = sc
}

func (s *Stan) Subsribes() {
	eventWorker := workers.Stan{EventStore}
	s.client.Subscribe("user_created", func(m *stan.Msg) {
		log.Printf("received message %v", m)
		eventWorker.HandleUserCreated(m)
	}, stan.DurableName("durable-wallet"))

	s.client.Subscribe("promotion_applied", func(m *stan.Msg) {
		log.Printf("received message %v", m)
		eventWorker.HandlePromotionApplied(m)
	}, stan.DurableName("durable-wallet"))
}
